src/3DFactory.R

```diff
@@ -0,0 +1,138 @@
+# LT 17/05/2022
+
+# prototype: 3D Factory AND Adjudicator are in the same source file to avoid
+# implementation of the interface between them
+
+# 3D Factory component
+
+# the 3D Factory generates a list of 3D regions and calculates their intolerance score
+
+# First prototype
+# creates 3D regions of variable length using upper bound of CI
+# each region is built by growing a sphere centered on the carbon-alpha of
+# a specific residue until the whole protein is included
+# a criteria (based on the upper bound) is used to decide the length of the region
+
+# file names and parameters are hard-coded for this prototype
+
+source("constraint_utils.R")
+
+## calculate upper bound of the o/e confidence interval
+get_oe_ci_up <- function(observed, expected, alpha = 0.05) {
+  return(qchisq((1 - alpha / 2), 2 * (observed + 1)) / 2 / expected)
+}
+
+## calculate intolerance metric of a region
+get_region_metric <- function(i_region, alpha, regions, oe_data) {
+  # indices of the residues belonging to this region
+  this_region <- regions[seq_len(i_region)]
+
+  # total number of missense variants observed in the region
+  total_obs <- sum(oe_data$obs[this_region])
+
+  # total number of missense variants expected in the region
+  total_exp <- sum(oe_data$exp[this_region])
+
+  # upper bound of o/e
+  oe_upper <- get_oe_ci_up(total_obs, total_exp, alpha)
+
+  c(obs = total_obs, exp = total_exp, oe = total_obs / total_exp, oe_upper = oe_upper)
+}
+
+### main
+## inputs
+# protein structure file
+prot_file <- "../structures/AF-P68133-F1-model_v2.pdb"
+
+# o/e file, out of the GenomicSimulator or a genomic database
+oe_file <- "../structures/AF-P68133-F1-model_v2.simmut.tsv"
+
+oe <- read.table(
+  file = oe_file, sep = "\t",
+  col.names = c("residue_index", "obs", "exp")
+)
+
+# calculate pairwise distances between residues based on their carbon alpha
+dist_mat <- get_dist_mat(prot_file)
+
+# number of residues in protein
+n <- nrow(dist_mat)
+
+##################
+# 3D Factory
+
+# get 3D region and its metrics for each residue
+regions_list <- lapply(seq_len(n), function(i_residue) {
+
+  # get the corresponding row of the distance matrix
+  row <- dist_mat[i_residue, ]
+
+  # grow the sphere = order by increasing distance
+  this_aa_regions <- seq_len(n)[order(row)]

```

Comment:
Again, I’d  choose a variable name without ‘this’

Reply:
Agreed. Changed to `current_sphere_regions`